File: src/rust/src/lib/interpreter.rs
use std::fs::File;
use std::io::{prelude::*, BufReader};

use crate::lib::config::*;

pub struct Interpreter<'a> {
    config: Config<'a>,
    port: isize,
    channel: isize
}

impl<'a> Interpreter<'a> {
    pub fn new(c: Config<'a>) -> Interpreter<'a> {
        Interpreter {
            config: c,
            port: -1,
            channel: -1
        }
    }

    pub fn run_file(&mut self, path: &str) {
        let file: File = File::open(path).unwrap();
        let reader = BufReader::new(file);

        for line in reader.lines() {
            let l = line.as_ref().unwrap().trim();
            if l != "" {
                self.run(l);
            }
            
        }
    }

    pub fn run(&mut self, command: &str) {
        let mut data = command.split(" ");
        
        match data.next().expect("Missing command") {
            "set" => match data.next().expect("Missing property to set") {
                "synth" => {
                    self.config.load_synth(data.next().expect("Missing synth id"));
                }
                "channel" => {
                    self.channel = data.next().expect("Missing channel number").parse::<isize>().expect("Failed to parse channel number");
                }
                _ => println!("Unknown property")
            }

            "midi" => (),

            "sysex" => {
                let alias = data.next().expect("Missing command alias");
                if self.config.aliases.contains_key(alias) {
                    let mut p: Vec<usize> = Vec::new();
                    while let Some(arg) = data.next() {
                        p.push(arg.parse::<usize>().expect("Could not parse argument value"));
                    }
                    self.midi(alias, p);
                } else {
                    println!("Alias not found");
                }
            }

            _ => ()
        }
    }

    pub fn midi(&self, command: &str, a: Vec<usize>) {
        let mut args = a;
        if self.channel > -1 {args.insert(0, self.channel as usize)}
        let c: String = self.config.aliases.get(command).expect("Command not found").build_vec(args);
        println!("{}", c);
    }
}

File: src/cpp/midi_prog/interpreter.cpp
#include "interpreter.h"
#include <iostream>
#include <fstream>
#include "utils.h"

Interpreter::Interpreter()
{

}

void Interpreter::run_file(std::string path)
{
    this->current_folder = get_folder(path);
    std::ifstream input( path );
    for(std::string line; getline(input, line); )
    {
        this->run(line);
    }
}

void Interpreter::run(std::vector<std::string> commands)
{
    for(auto c : commands)
    {
        this->run(c);
    }
}

void Interpreter::run(std::string command)
{
    trim(command);
    if(command == "") return;

    std::vector<std::string> tokens = split(command, ' ', true);
    std::string c = tokens[0];

    //std::cout << "Running "<< c << std::endl;

    if(c == "synth")
    {
        this->sysex = this->load_synth(tokens[1]);
    }
    else if(c == "interactive")
    {
        this->interactive();
    }
    else if(c == "config")
    {
        std::string path = (this->folder != "" ? this->folder + "/" : "") + tokens[1];
        this->config.run_file(path);
    }
    else if(c == "midiconfig")
    {
        std::string path = (this->folder != "" ? this->folder + "/" : "") + tokens[1];
        this->config.run_file(path);
        if(this->config.get_synth("midi") != nullptr)
        {
            this->midi = this->load_synth("midi");
        }
        else
        {
            std::cout << "Could not load the generic midi configuration" << std::endl;
        }

    }
    else if(c == "port")
    {
        this->port = std::stoi(tokens[1]);
        this->interface.setPort(this->port);
    }
    else if(c == "portlist")
    {
        this->interface.listPorts();
    }
    else if(c == "channel")
    {
        this->channel = std::stoi(tokens[1]);
    }
    else if(c == "receive")
    {
        this->interface.waitMidi();
    }
    else if(c == "send")
    {
        std::vector<unsigned char> data = stob(tokens[1]);
        this->interface.send(data);
    }
    else if(c == "sysex")
    {
        std::string alias = tokens[1];
        std::vector<long> values;

        if(this->channel >= 0) values.push_back(channel);

        for(size_t i=2; i<tokens.size(); i++)
        {
            values.push_back(std::stoi(tokens[i]));
        }

        std::vector<unsigned char> data = this->sysex[alias]->generateBytes(values);

        std::cout << this->sysex[alias]->generate(values) << std::endl;
        std::cout << std::hex;
        for(auto d : data)
        {
            std::cout << std::hex << static_cast<int>(d) << " ";
        }
        std::cout << std::dec << std::endl;

        this->interface.send(data);
    }
    else if(c == "midi")
    {
        std::string alias = tokens[1];
        std::vector<long> values;

        if(this->channel >= 0) values.push_back(channel);

        for(size_t i=2; i<tokens.size(); i++)
        {
            values.push_back(std::stoi(tokens[i]));
        }

        std::vector<unsigned char> data = this->midi[alias]->generateBytes(values);

        std::cout << this->midi[alias]->generate(values) << std::endl;
        std::cout << std::hex;
        for(auto d : data)
        {
            std::cout << std::hex << static_cast<int>(d) << " ";
        }
        std::cout << std::dec << std::endl;

        this->interface.send(data);
    }
    else if(c == "folder")
    {
        if(tokens[1] == "relative")
        {
            this->folder = this->current_folder;
        }
        else if(tokens[1] == "absolute")
        {
            this->folder = "";
        }
    }
    else
    {
        std::cout << "Unknown command : " << c << std::endl;
    }
}

void Interpreter::interactive()
{
    std::string command;
    while (true)
    {
        if(this->synth != nullptr) std::cout << this->synth->getName();
        std::cout << "@" << this->port;
        if(this->channel > -1) std::cout << "-" << this->channel;
        std::cout << " > ";

        std::getline(std::cin, command);

        std::vector<std::string> commands = split(command, '|', true);

        if(command == "exit"){
            break;
        } else{
            this->run(commands);
        }
    }
}

std::map<std::string, MIDICommand*> Interpreter::load_synth(std::string id)
{
    Synth *s = config.get_synth(id);
    this->synth = s;

    std::vector<MIDICommand> *commands = s->getCommands();
    std::map<std::string, MIDICommand*> midi;

    for(size_t i=0; i<commands->size(); i++)
    {
        MIDICommand *m = &commands->at(i);
        for(auto a : m->getAliases())
        {
            std::pair<std::string, MIDICommand*> p(a,m);
            midi.insert(p);
        }
    }

    return midi;
}
